File: src/ba01/ba01k_compute_freq/ba01k.js
/*
Rosalind: BA1K
Generate the Frequency Array of a String

Given an integer k, we define the frequency array of a string Text as an array
of length 4k, where the i-th element of the array holds the number of times that
the i-th k-mer (in the lexicographic order) appears in Text (see Figure 1.

kmer      AA  AC  AG  AT  CA  CC  CG  CT  GA  GC  GG  GT  TA  TC  TG  TT
index      0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15
frequency  3   0   2   0   1   0   0   0   0   1   3   1   0   0   1   0

Computing a Frequency Array
Generate the frequency array of a DNA string.

Given: A DNA string Text and an integer k.

Return: The frequency array of k-mers in Text.

Sample Dataset
ACGCGGCTCTGAAA
2

Sample Output
2 1 0 0 0 0 2 2 1 2 1 0 0 1 1 0

═════════════════════════════════════════════════

    [ Where am I? ]

    * Where in the genome does DNA replication begin? (Chapter 01)
      -> Locating replication origin (OriC reion - DnaA box)
      -> minimum skewness (BA1F)
      -> Finding hidden message (DnaA box) in OriC
      ↓
    * Most frequent words problem
      -> count words (BA1A)
      -> find frequent words in a string (BA1B)
      -> find all occurrence of a pattern in a string (BA1D)
      -> Clump finidng problem (BA1E)
      ↓
    * Most frequent words with mismatch problem
      -> hamming distance (BA1G)
      -> Find All Approximate Occurrences of a Pattern in a String (BA1H)
      -> Find the Most Frequent Words with Mismatches in a String (BA1I) - by frequency array
        -> HERE: frequency array
          -> PREV: number to pattern (BA1M), pattern to number (BA1L)
        -> NEXT: neigbors (BA1N)

Plan 1.
- pseudocode:
  ╔═══════════════════════════════════════════════════╗
  ║  COMPUTINGFREQUENCIES(Text, k)                    ║
  ║    for i <- 0 to 4^k- 1                           ║
  ║      FREQUENCYARRAY(i) <- 0                       ║
  ║    for i <- 0 to |Text| - k                       ║
  ║      Pattern <- Text(i, k)                        ║
  ║      j <- PATTERNTONUMBER(Pattern)                ║
  ║      FREQUENCYARRAY(j) <- FREQUENCYARRAY(j) + 1   ║
  ║    return FREQUENCYARRAY                          ║
  ╚═══════════════════════════════════════════════════╝

*/

// BA01M: number to pattern
function number_to_pattern(num, k) {
    /**
     * (int,int) -> str
     * returns a pattern from a number
     * number_to_pattern(15, 2) == "TT"
     */
    pattern = ""
    for (let i = 0; i < k; i++) {
        temp = Math.floor(num / Math.pow(4, (k - i - 1)));
        if (temp == 0) pattern += 'A';
        else if (temp == 1) pattern += 'C';
        else if (temp == 2) pattern += 'G';
        else if (temp == 3) pattern += 'T';
        num -= temp * Math.pow(4, k - i - 1);
    }
    return pattern
}

// BA01L: number to pattern, recursive
function number_to_pattern_rec(num, k) {
    /**
     * (int,int) -> str
     * returns a pattern from a number
     * number_to_pattern_rec(15, 2) == "TT"
     */
    const dic = { 0: "A", 1: "C", 2: "G", 3: "T" };
    if (k == 0) return ""
    else {
        temp = Math.floor(num / Math.pow(4, (k - 1)));
        num -= temp * Math.pow(4, k - 1);
        return dic[temp] + number_to_pattern(num, k - 1)
    }
}

// BA01L: pattern to number
function pattern_to_number(pattern) {
    /**
     * str -> int
     * returns a number from a pattern
     * pattern_to_number("TT") == 15
     */
    const dic = { 'A': 0, 'C': 1, 'G': 2, 'T': 3 };
    let number = 0;
    for (let i = 0; i < pattern.length; i++) {
        let pw = pattern.length - i - 1;
        let nuc = 0;
        let chr = pattern[i];
        if (chr == 'A') nuc = 0;
        else if (chr == 'C') nuc = 1;
        else if (chr == 'G') nuc = 2;
        else if (chr == 'T') nuc = 3;
        let pw_val = Math.pow(4, pw);
        number += pw_val * nuc;
    }
    return number;
}

// BA01L: pattern to number, recursive
function pattern_to_number_rec(pattern) {
    /**
     * str -> int
     * returns a number from a pattern
     * pattern_to_number_rec("TT") == 15
     */
    const dic = { 'A': 0, 'C': 1, 'G': 2, 'T': 3 };
    if (pattern.length == 0) {
        return 0;
    } else {
        return dic[pattern[0]] * Math.pow(4, pattern.length - 1) + pattern_to_number(pattern.slice(1));
    }
}

// BA01K: compute frequency
function compute_freq(text, k) {
    /**
     * (str,int) -> [int]
     * returns a frequency array
     * compute_freq('ACGCGGCTCTGAAA', 2)
     * == [2,1,0,0,0,0,2,2,1,2,1,0,0,1,1,0]
     */
    let freq_arr = new Array(Math.pow(4, k));  // TODO: find the fastest way
    for (let i = 0; i < Math.pow(4, k); i++) freq_arr[i] = 0;
    for (let i = 0; i < text.length - k + 1; i++) {
        let kmer = text.slice(i, i + k);
        let num = pattern_to_number(kmer);
        freq_arr[num] += 1;
    }
    return freq_arr;
}

// BA01K: compute frequency
function compute_freq_dic(text, k) {
    /**
     * (str,int) -> ({int:int},int)
     * returns a frequency array
     */
    let freq_dic = {};
    let max_cnt = 0;
    for (let i = 0; i < text.length - k + 1; i++) {
        let kmer = text.slice(i, i + k);
        let num = pattern_to_number(kmer);
        if (!(num in freq_dic)) freq_dic[num] = 1;
        else {
            freq_dic[num] += 1;
            if (freq_dic[num] > max_cnt) max_cnt = freq_dic[num];
        }
    }
    return [freq_dic, max_cnt];
}

// main
function main() {
    const fs = require('fs');
    let lines = fs.readFileSync('/home/wsl/rosalind/data/ba01k.txt').toString().split("\n");
    const pattern = lines[0]
    const k = parseInt(lines[1]);

    let startTime = performance.now()
    let answer = "";
    for (const elem of compute_freq(pattern, k)) {
        answer += elem + " ";
    }
    console.log(answer);
    console.log(`${performance.now() - startTime} milliseconds`)
}

// execute main function
main()

File: src/ba01/ba01k_compute_freq/ba01k.cpp
/*
Rosalind: BA1K
Generate the Frequency Array of a String

Given an integer k, we define the frequency array of a string Text as an array
of length 4k, where the i-th element of the array holds the number of times that
the i-th k-mer (in the lexicographic order) appears in Text (see Figure 1.

kmer      AA  AC  AG  AT  CA  CC  CG  CT  GA  GC  GG  GT  TA  TC  TG  TT
index      0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15
frequency  3   0   2   0   1   0   0   0   0   1   3   1   0   0   1   0

Computing a Frequency Array
Generate the frequency array of a DNA string.

Given: A DNA string Text and an integer k.

Return: The frequency array of k-mers in Text.

Sample Dataset
ACGCGGCTCTGAAA
2

Sample Output
2 1 0 0 0 0 2 2 1 2 1 0 0 1 1 0

═════════════════════════════════════════════════
Pseudocode:
  ╔═══════════════════════════════════════════════════╗
  ║  COMPUTINGFREQUENCIES(Text, k)                    ║
  ║    for i <- 0 to 4^k- 1                           ║
  ║      FREQUENCYARRAY(i) <- 0                       ║
  ║    for i <- 0 to |Text| - k                       ║
  ║      Pattern <- Text(i, k)                        ║
  ║      j <- PATTERNTONUMBER(Pattern)                ║
  ║      FREQUENCYARRAY(j) <- FREQUENCYARRAY(j) + 1   ║
  ║    return FREQUENCYARRAY                          ║
  ╚═══════════════════════════════════════════════════╝

═════════════════════════════════════════════════
Plan 1.
  (1) get all k-mer patterns: permutation with repetition
  (2) sort k-mer patterns above
  (3) define a function for calculating frequency
      - 'std::count'
  (4) get frequency array

Plan 2.
  (1) define helper functions:
      - number to pattern
      - pattern to number
  (2) create an empty array with its size 4^n
  (3) figure out the relationship between the index and the pattern

Topic:
-
*/

#include <cmath>
#include "/home/wsl/rosalind/include/utils.hpp"

/************************************************
function prototypes
************************************************/

template <typename T>
void print_vec(const T& t);

int get_num_lines(std::string filename);
std::vector<std::string> get_lines(std::string filename);
std::vector<std::string> split_str(std::string str, std::string delimiter);
std::vector<std::string> perm_with_repetition(std::vector<std::string> vec,
    std::string nuc, int num);
int get_freq(const std::string&, const std::string&);
std::vector<int> compute_freq_brute(std::string, int);

long long int pattern_to_number(std::string);
std::string number_to_pattern(int, int);
std::vector<int> compute_freq(std::string, int);

/************************************************
main
************************************************/

int main(int argc, const char* argv[]) {
    const std::string filename = argc < 2 ? "/home/wsl/rosalind/data/ba01k.txt" : argv[1];
    std::vector<std::string> lines = read_lines(filename);
    std::string pattern = lines[0];
    int num_k = std::stoi(lines[1]);

    std::vector<int> freq_vec = compute_freq(pattern, num_k);
    print_vec(freq_vec);

    return 0;
}

/************************************************
function definitions
************************************************/

std::vector<std::string> perm_with_repetition(std::vector<std::string> vec,
    std::string nuc, int num) {
    if (num == 0) {
        return vec;
    }
    else if (vec.size() == 0) {
        return perm_with_repetition({ "A", "C", "G", "T" }, nuc, num - 1);
    }
    else {
        std::vector<std::string> vec_new;
        for (auto elem : vec) {
            for (auto& chr : nuc) {
                std::string temp = elem + chr;
                vec_new.push_back(temp);
            }
        }
        vec = std::move(vec_new);
        return perm_with_repetition(vec, nuc, num - 1);
    }
}

int get_freq(const std::string& genome, const std::string& pattern) {
    int cnt = 0;
    for (int i = 0; i < genome.size() - pattern.size() + 1; i++) {
        if (pattern == genome.substr(i, pattern.size())) cnt++;
    }
    return cnt;
}

// BA01K: compute frequency, brute
std::vector<int> compute_freq_brute(std::string genome, int num_k) {
    std::vector<std::string> perm = perm_with_repetition({}, "ACGT", num_k);
    std::vector<int> freq_vec;  // TODO: problem here
    for (auto kmer : perm) {
        freq_vec.push_back(get_freq(genome, kmer));
    }
    return freq_vec;
}

// BA01L: pattern to number
long long int pattern_to_number(std::string pattern) {
    long long int number = 0;
    int len = pattern.size();
    for (int i = 0; i < len; i++) {
        int pw = len - i - 1;
        int nuc;
        switch (pattern[i]) {
        case 'A':
            nuc = 0;
            break;
        case 'C':
            nuc = 1;
            break;
        case 'G':
            nuc = 2;
            break;
        case 'T':
            nuc = 3;
            break;
        }
        long double pw_val = std::pow(4.0, pw * 1.0);
        number += (pw_val * nuc);
    }
    return number;
}

// BA01M: number to pattern
std::string number_to_pattern(int number, int num_k) {
    std::string pattern;
    for (int i = 0; i < num_k; i++) {
        int temp = number / std::pow(4, num_k - i - 1);
        switch (temp) {
        case 0:
            pattern += 'A';
            break;
        case 1:
            pattern += 'C';
            break;
        case 2:
            pattern += 'G';
            break;
        case 3:
            pattern += 'T';
            break;
        }
        number -= temp * std::pow(4, num_k - i - 1);
    }
    return pattern;
}

// BA01K: compute frequency
std::vector<int> compute_freq(std::string pattern, int num_k) {
    std::vector<int> freq_vec(pow(4, num_k), 0);
    for (int i = 0; i < pattern.size() - num_k + 1; i++) {
        std::string kmer = pattern.substr(i, num_k);
        int temp = pattern_to_number(kmer);
        freq_vec[temp] += 1;
    }
    return freq_vec;
}